seminare/users/mixins/permissions.py

```diff
@@ -0,0 +1,60 @@
+from typing import Callable, Protocol
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, PermissionDenied
+from django.http import HttpRequest, HttpResponseRedirect
+
+from seminare.users.logic.permissions import has_contest_role
+from seminare.users.models import ContestRole, User
+
+
+class MixinProtocol(Protocol):
+    request: HttpRequest
+    dispatch: Callable
+
+
+class AccessMixin(MixinProtocol):
+    def check_access(self) -> bool:
+        raise ImproperlyConfigured(
+            f"{self.__class__.__name__} does not override check_access method."
+        )
+
+    def dispatch(self, request: HttpRequest, *args, **kwargs):
+        if not request.user.is_authenticated:
+            return HttpResponseRedirect(settings.LOGIN_URL)
+
+        if not self.check_access():
+            raise PermissionDenied()
+
+        return super().dispatch(request, *args, **kwargs)
+
+
+class ContestAccessMixin(AccessMixin):
+    required_role = None
+
+    def get_permission_contest(self):
+        raise ImproperlyConfigured(
+            f"{self.__class__.__name__} does not provide permission contest."
+        )
```

Comment:
How is this going to be used? Isn't it better to be compatible with `WithContest` Mixin?